bluefog/common/mpi_controller.cc

```diff
@@ -432,10 +470,75 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
             "output for details.");
       }
     } else {
-      error_message = mpi_ctx_.NeighborValueExchangeWithConstantElements(
-        sendbuf, (void *)entry.output->data(), num_elements, entry.output->dtype(),
-        entry.send_neighbors.get(), entry.recv_neighbors.get()
-      );
+      int nsend = entry.send_neighbors->size();
+      int nrecv = entry.recv_neighbors->size();
+
+      // Ensure the lifecycle of the weighted tensors are alive after communication.
+      std::vector<std::unique_ptr<common::Tensor>> weighted_tensors;
+      if (entry.dst_weighting_enabled) {
+        for (int i = 0; i < nsend; ++i) {
+          auto weighted_tensor_ptr = entry.tensor->data_weight(entry.send_weights->at(i));
+          weighted_tensors.push_back(std::move(weighted_tensor_ptr));
+        }
+      }
+      std::shared_ptr<common::ReadyEvent> ready_event =
+          entry.context->RecordReadyEvent(entry.device);
+
+      std::vector<MPI_Request> requests(nsend + nrecv);
+      std::vector<MPI_Status> statuses(nsend + nrecv);
+      int element_size = mpi_ctx_.GetMPITypeSize(entry.output->dtype());
+      for (int i = 0; i < nrecv; ++i) {
+        void* recvbuf = (void*)(static_cast<const char*>(entry.output->data()) +
+                                num_elements * i * element_size);
+        int ret_code = MPI_Irecv(
+            recvbuf, num_elements, mpi_ctx_.GetMPIDataType(entry.output),
+            entry.recv_neighbors->at(i),
+            mpi_ctx_.rank_ + entry.recv_neighbors->at(i),
+            mpi_ctx_.GetMPICommunicator(Communicator::GRAPH),
+            &requests[i + nsend]);
+        if (ret_code != MPI_SUCCESS) {
+          throw std::runtime_error(
+              "MPI_Irecv (for dynamic neighbor_allreduce) failed, see MPI "
+              "output for details.");
+        }
```

Comment:
It is my function, we should use the macro `MPICHECK`, which is defined in common.h

Reply:
Done.